Fix oauth using flask-dance

emilyfeder opened the pull request:
**Overview**
Using the flask-dance library to handle oauth.
This worked on my machine so that each new incognito window did indeed ask you to re-auth! Hopefully it works on yours! Note the changes to the `environment.yml` file, so you'll need to make sure your dependencies match (i.e. versions of `requests`, `oauthlib`, `requests-oauthlib` changed -- I deleted our own dependencies on them and then let the install of `flask-dance` handling installing the proper versions).

I moved everything in `renxt_client.py` out of a class and just made them all module functions, since nothing needs to be initialized; `renxt` is now a variable that will always represent a session between our app and the sky api that has a different token per browser session. That happens due to some [magic stuff](https://github.com/jcua-tech/jcua-renxt/compare/ef%2Ffix-flask-dance?expand=1#diff-8029191396861061fb9c713fa182ea5b) I don't quite understand.

Still to fix:
- The secret key setting in `jcua_renxt/__init__.py` should obviously not be done in the code itself
- Not sure if this handles token refreshing. I did not have time to look into it
- Oh, and the tests probably broke somehow

alexsoble commented:
Taking a look at this now!

alexsoble commented:
@emilyfeder Working on a PR to move the secret key to `app_config`.

alexsoble commented:
@emilyfeder Verified that this works for me as expected (handles authorization) in development mode, and that the sessions issue from earlier has been resolved! Excellent!! 🎉 🎉 

What type of review would be most helpful for this PR? Are there particular areas of the code you'd want me to look at, questions that you're wondering about, or PRs into this PR that would be helpful?

emrapport commented:
This all looks good to me! I tried loading the page in incognito mode, and in normal mode after clearing my cookies, and I got the reauthorization page both times. Thanks for explaining what you've learned about oauth and flask-dance; I think the changes we've made make sense.

alexsoble commented:
@emilyfeder Taking a quick 👀 at the test failures, it looks like these are mostly mechanical changes needed since RenxtClient isn't a class anymore. Happy to work on fixing those up and open a PR into this PR if that would be helpful! 😄

emilyfeder commented:
Alex, sorry I missed your messages over the last couple of hours. I don't have any worries that come to mind about the code. In the past hours, I moved the secret to the app_config and fixed the existing tests. The part I couldn't figure out was how to get an authed renxt client as a test fixture. However, I do think if we used a test_client to actually make requests in tests we could mock the responses as demo'd here: https://flask-dance.readthedocs.io/en/latest/testing.html
Existing tests are only for the AddressGeocoder which pass in a mocked renxt_client, so we don't need to deal with that yet.

alexsoble commented on README.md:
````diff
@@ -22,6 +22,12 @@ Make sure to keep environment.yml up to date with any dependencies you install:
 conda env export --file environment.yml
 ```
 
+If someone else has updated the environment.yml, then update your environment using:
````
Great README addition, I was looking for this myself earlier today.

alexsoble commented on jcua_renxt/__init__.py:
```diff
@@ -1,11 +1,16 @@
 from flask import Flask
+from jcua_renxt import app_config
 from jcua_renxt.blueprints.hello_world import hello_world
+from jcua_renxt.blueprints.renxt_oauth import renxt_bp
```
Tiny tiny thing, but if you're open to it, I would suggest using `renxt_blueprint` instead of `renxt_bp` — folks without a Python/Flask background might not know what "bp" stands for, and one or two-letter variable names can be vague.

alexsoble commented on jcua_renxt/app_config.py:
```diff
@@ -38,8 +41,9 @@
 
 elif env == 'test':
-    OAUTH_CALLBACK_URL = 'fake_auth_callback'
+    TESTING = True #for flask testing
```
It looks like this ENV variable is never used, can we remove it? Introducing "testing" flags to auth-related code is tricky because you want to be super super sure you're confident about then the "testing" flag is/isn't triggered.

emilyfeder replied:
Ah yes, at one point I thought this was getting passed implicitly into create_app, but I didn't end up doing that. I'll set it in the test fixture itself.

alexsoble commented on tests/test_renext_client.py:
```diff
@@ -1,11 +1,14 @@
-from jcua_renxt import renxt_client
+from jcua_renxt import renxt_client as renxt
```
I love how changing this one line let you keep the rest of the file unchanged even though `renxt_client.RenxtClient()` is gone! ✨

alexsoble commented on environment.yml:
```diff
@@ -45,11 +42,16 @@ dependencies:
     - pytest==3.7.1
     - ratelim==0.1.6
-
+    - requests==2.21.0
```
I see we're now using more up-to-date versions of all these 🔑 dependencies, which is a good thing.
